In [ ]:
# Update sklearn to prevent version mismatches
#!pip install sklearn --upgrade

In [ ]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
#!pip install joblib

In [2]:
import pandas as pd
import numpy as np

Import our white and red wine data from the csvs

In [3]:
red_wine_df = pd.read_csv("winequality-red.csv", sep=';')
white_wine_df = pd.read_csv("winequality-white.csv", sep=';')
print(white_wine_df.count())
print(red_wine_df.count())

fixed acidity           4898
volatile acidity        4898
citric acid             4898
residual sugar          4898
chlorides               4898
free sulfur dioxide     4898
total sulfur dioxide    4898
density                 4898
pH                      4898
sulphates               4898
alcohol                 4898
quality                 4898
dtype: int64
fixed acidity           1599
volatile acidity        1599
citric acid             1599
residual sugar          1599
chlorides               1599
free sulfur dioxide     1599
total sulfur dioxide    1599
density                 1599
pH                      1599
sulphates               1599
alcohol                 1599
quality                 1599
dtype: int64


## Balanced (Normalized) Data for White Wine Data (1599 data points) and Red Wine Data (1599 data points) 

In [4]:
#take a random sample of white wines to match the red wine data
random_white_df = white_wine_df.sample(1599,random_state=54).reset_index(drop=True)

Only balanced values are baseline condition we can use:

In [6]:
#if we want our only features of the model to have balanced white wine and red wine data, we use this data frame:
baseline_red_df = red_wine_df.copy()
baseline_white_df = random_white_df.copy()
only_balanced_df = pd.concat([baseline_red_df, baseline_white_df])
only_balanced_df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.9              0.18         0.30             2.0      0.038   
1595            7.8              0.25         0.34            13.7      0.044   
1596            6.5              0.32         0.34             5.7      0.044   
1597            7.0              0.46         0.39             6.2      0.039   
1598            7.8              0.26         0.44             1.3      0.037   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 39.0                 190.0  0.99140  3.32       0.37   
1595                 66.0                 184.0  0.99976  3.22       0.75   
1596                 27.0                  91.0  0.99184  3.28       0.60   
1597                 46.0                 163.0  0.99280  3.21       0.35   
1598                 43.0                 132.0  0.99440  3.18       0.65   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     12.2        6  
1595      8.9        5  
1596     12.0        7  
1597     12.2        7  
1598     10.0        5  

[3198 rows x 12 columns]

In [ ]:
#if we want to include wine colors as features in our model:
random_white_df["color"] = "white"
red_wine_df["color"] = "red"
red_wine_df

Added Color with get dummies (Or in case we want to remove it)

Add buckets to given range of quality values

Choose range of quality values 

### Moved quality values into buckets

# Unbalanced (Denormalized) Data for White Wine Data (4898 data points) and Red Wine Data (1599 data points) 

# Read the CSV and Perform Basic Data Cleaning

In [ ]:
df1 = pd.read_csv("winequality-white.csv", sep=';')
df2 = pd.read_csv("winequality-red.csv", sep=';')
combinedf = pd.concat([df1,df2])
# combinedf = combinedf.drop(columns=['density'])
# combinedf.drop
# combinedf = combinedf.dropna(axis='columns', how='all')
# Drop the null rows
# combinedf = combinedf.dropna()
# combinedf.head()

In [ ]:
combinedf = combinedf.loc[\
(combinedf["quality"] < 9 ) & \
(combinedf["quality"] > 3 )  \
] 

In [ ]:
combinedf.groupby('quality').count()

# Select your features (columns)

In [ ]:
# Set features. Drop quality, y value. 
X = combinedf.drop(columns='quality')
X.head()

# Create a Train Test Split

Use `quality` for the y values

In [ ]:
from sklearn.model_selection import train_test_split
#80% train, 20% test. y is this one column
y = combinedf['quality']
#random state 42 will have same picks for x test and y test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [ ]:
X_train.head()

# Train the Model using Random Forest 



In [ ]:
#Train the model
from sklearn.ensemble import RandomForestClassifier
rfc=RandomForestClassifier()
rfc.fit(X_train, y_train)

In [ ]:
#Test the model
from sklearn.metrics import confusion_matrix, classification_report
predictions = rfc.predict(X_test)
print(classification_report(y_test, predictions))

# Save the Model

In [ ]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'rfcredwhitewine567.sav'
joblib.dump(predictions, filename)